base_layer/core/src/base_node/sync/header_sync/synchronizer.rs

```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
                 ));
             }
 
-            let steps_back = resp.fork_hash_index.saturating_add(offset as u64);
-            return Ok((resp, block_hashes, steps_back));
+            let reorg_steps_back = resp.fork_hash_index.saturating_add(offset as u64);
+            let proto::FindChainSplitResponse {
+                headers,
+                fork_hash_index,
+                tip_height: remote_tip_height,
+            } = resp;
+            return Ok(PeerChainSplitResponse {
+                block_hashes,
+                reorg_steps_back,
+                headers,
+                fork_hash_index,
+                remote_tip_height,
+            });
         }
     }
 
     /// Attempt to determine the point at which the remote and local chain diverge, returning the relevant information
-    /// of the chain split (see [SyncStatus]).
+    /// of the chain split (see [HeaderSyncStatus]).
     ///
-    /// If the local node is behind the remote chain (i.e. `SyncStatus::Lagging`), the appropriate `ChainSplitInfo` is
-    /// returned, the header validator is initialized and the preliminary headers are validated.
+    /// If the local node is behind the remote chain (i.e. `HeaderSyncStatus::Lagging`), the appropriate
+    /// `ChainSplitInfo` is returned, the header validator is initialized and the preliminary headers are validated.
     async fn determine_sync_status(
         &mut self,
-        sync_peer: &SyncPeer,
-        local_tip_header: ChainHeader,
-        client: &mut rpc::BaseNodeSyncRpcClient,
-    ) -> Result<SyncStatus, BlockHeaderSyncError> {
-        let (resp, block_hashes, steps_back) = self
-            .find_chain_split(sync_peer.node_id(), client, NUM_INITIAL_HEADERS_TO_REQUEST as u64)
-            .await?;
-        let proto::FindChainSplitResponse {
-            headers,
-            fork_hash_index,
-            tip_height: remote_tip_height,
-        } = resp;
-
-        if steps_back > 0 {
+        sync_peer: &str,
+        best_header: ChainHeader,
+        best_block: ChainHeader,
+        peer_response: PeerChainSplitResponse,
+    ) -> Result<HeaderSyncStatus, BlockHeaderSyncError> {
+        if peer_response.reorg_steps_back > 0 {
             debug!(
                 target: LOG_TARGET,
                 "Found chain split {} blocks back, received {} headers from peer `{}`",
-                steps_back,
-                headers.len(),
+                peer_response.reorg_steps_back,
+                peer_response.headers.len(),
                 sync_peer
             );
         }
 
-        // If the peer returned no new headers, this means header sync is done.
-        if headers.is_empty() {
-            if fork_hash_index > 0 {
+        // If the peer returned no new headers, they have no headers, but may still have more blocks than we have,
+        // thus have a higher accumulated difficulty.
```

Comment:
This comment makes no sense. The headers are what determines the accumulated difficulty.

Reply:
Fixed